vendor/github.com/danwinship/knftables/util.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2023 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package knftables
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+// PtrTo can be used to fill in optional field values in objects
+func PtrTo[T any](val T) *T {
+	return &val
+}
+
+var numericPriorities = map[string]int{
+	"raw":      -300,
+	"mangle":   -150,
+	"dstnat":   -100,
+	"filter":   0,
+	"security": 50,
+	"srcnat":   100,
+}
+
+var bridgeNumericPriorities = map[string]int{
+	"dstnat": -300,
+	"filter": -200,
+	"out":    100,
+	"srcnat": 300,
+}
+
+// ParsePriority tries to convert the string form of a chain priority into a number
+func ParsePriority(family Family, priority string) (int, error) {
+	val, err := strconv.Atoi(priority)
+	if err == nil {
+		return val, nil
+	}
+
+	modVal := 0
+	if i := strings.IndexAny(priority, "+-"); i != -1 {
+		mod := priority[i:]
+		modVal, err = strconv.Atoi(mod)
+		if err != nil {
+			return 0, fmt.Errorf("could not parse modifier %q: %w", mod, err)
+		}
+		priority = priority[:i]
+	}
+
+	var found bool
+	if family == BridgeFamily {
+		val, found = bridgeNumericPriorities[priority]
+	} else {
+		val, found = numericPriorities[priority]
+	}
+	if !found {
+		return 0, fmt.Errorf("unknown priority %q", priority)
+	}
+
+	return val + modVal, nil
+}
+
+// Concat is a helper (primarily) for constructing Rule objects. It takes a series of
+// arguments and concatenates them together into a single string with spaces between the
+// arguments. Strings are output as-is, string arrays are output element by element,
+// numbers are output as with `fmt.Sprintf("%d")`, and all other types are output as with
+// `fmt.Sprintf("%s")`. To help with set/map lookup syntax, an argument of "@" will not
+// be followed by a space, so you can do, eg, `Concat("ip saddr", "@", setName)`.
+func Concat(args ...interface{}) string {
```

Comment:
So this is all basically copied from `utilproxy.LineBuffer` of course...

I keep thinking about how we can optimize knftables's memory usage in the face of large transactions, the way that the iptables proxy (a) reuses the same buffers from sync to sync, and (b) reuses the same `[]string` throughout a single sync.

The buffer reusing ought to be easy to add: `realNFTables` could just keep a `bytes.Buffer` around rather than allocating a new one for each transaction. (This needs a little bit of thought to deal with the possibility of multiple pending transactions and stuff, but it's all solvable.)

The equivalent of the `[]string` reusing would be to optimize the way that `Concat()` works...

I realized at some point that if you don't care about reading back the `Rule string` after you construct a `knftables.Rule`, then instead concatenating the rule pieces when you construct the rule, we could just hold off doing it until the transaction gets written out. eg, you'd do something like

```
tx.Add(&knftables.Rule{
        Chain: "mychain",
        Rule: []interface{
                ipX, "daddr", myIP,
                "drop",
        },
})
```

and then the code that's currently in `Concat` would get moved to `Rule.writeOperation`, which would just write them directly into the transaction's `bytes.Buffer` without ever constructing a string containing all of them.

But this is a little uglier in the case where you're constructing a simple `Rule` with a single string... Then I thought maybe generics, and have `type Rule[string | io.Writer] struct { ... Rule T }`, but go doesn't infer generic types in struct constructors the way it does in function calls, so you'd have to explicitly say `&knftables.Rule[string]` or `&knftables.Rule[knftables.ConcatResult]` or whatever.

Maybe we could do

```
        Rule      string
        BuildRule []interface{}
```

and you just set one or the other? and have a `GetRule() string` to simplify the reading-back case...

Reply:
is sync.Pool an option https://pkg.go.dev/sync#example-Pool ?